src/components/Editor/DrawingBoard/Canvas/Container.ts

```diff
@@ -0,0 +1,239 @@
+    } else {
+      throw new Error(`Undefined tool: ${tool}`);
+    }
+  }
+
+  setMouseClass(tool: ToolType) {
+    this.upperWrapper.getCanvas().className = 'canvas canvas-upper';
+
+    if (tool === ToolType.Line || tool === ToolType.Rect) {
+      this.upperWrapper.getCanvas().classList.add('crosshair', 'canvas-touch-none');
+    } else if (tool === ToolType.Eraser) {
+      this.upperWrapper.getCanvas().classList.add('eraser', 'canvas-touch-none');
+    } else if (tool === ToolType.Selector) {
+      this.upperWrapper.getCanvas().classList.add('canvas-touch-none');
+    }
+  }
+
+  getPointFromTouchyEvent(evt: TouchyEvent): Point {
+    let originY;
+    let originX;
+    if (window.TouchEvent && evt instanceof TouchEvent) {
+      originY = evt.touches[0].clientY;
+      originX = evt.touches[0].clientX;
+    } else {
+      originY = evt.clientY;
+      originX = evt.clientX;
+    }
+    originY += window.scrollY;
+    originX += window.scrollX;
+    return {
+      y: originY - this.offsetY,
+      x: originX - this.offsetX,
+    };
+  }
+
+  onMouseDown(evt: TouchyEvent) {
+    touchy(this.upperWrapper.getCanvas(), addEvent, 'mousemove', this.onMouseMove);
+    this.dragStatus = DragStatus.Drag;
+
+    const point = this.getPointFromTouchyEvent(evt);
+
+    this.worker.mousedown(point, { color: this.color });
+  }
+
+  onMouseMove(evt: TouchyEvent) {
+    const point = this.getPointFromTouchyEvent(evt);
+    if (this.isOutside(point)) {
+      this.onMouseUp();
+      return;
+    }
+
+    if (this.dragStatus === DragStatus.Stop) {
+      return;
+    }
+
+    this.worker.mousemove(point);
+  }
+
+  onMouseUp() {
+    touchy(this.upperWrapper.getCanvas(), removeEvent, 'mousemove', this.onMouseMove);
+    this.dragStatus = DragStatus.Stop;
+
+    this.worker.mouseup();
+    this.emit('mouseup');
+  }
+
+  isOutside(point: Point): boolean {
+    if (
+      point.y < 0 ||
+      point.x < 0 ||
+      point.y > this.lowerWrapper.getHeight() ||
+      point.x > this.lowerWrapper.getWidth()
+    ) {
+      return true;
+    }
+    return false;
+  }
+
+  drawAll(shapes: Array<Shape>, wrapper: CanvasWrapper = this.lowerWrapper) {
+    this.clear(wrapper);
+    for (const shape of shapes) {
+      if (shape.type === 'line') {
+        drawLine(wrapper.getContext(), shape);
+      } else if (shape.type === 'eraser') {
+        drawLine(wrapper.getContext(), shape);
+      } else if (shape.type === 'rect') {
+        drawRect(wrapper.getContext(), shape);
+      }
+    }
+  }
+
+  clear(wrapper: CanvasWrapper = this.lowerWrapper) {
+    wrapper.clear();
+  }
+
+  emit(name: string, ...args: Array<unknown>) {
+    this.eventDispatcher.emit(name, ...args);
+  }
+
+  addEventListener(name: string, cb: Function) {
+    this.eventDispatcher.addEventListener(name, cb);
+  }
+
+  removeEventListener(name: string, cb?: Function) {
+    this.eventDispatcher.removeEventListener(name, cb);
+  }
+}
```

Comment:
I'm curious about why `emit, addEventListener , removeEventListener ` put in the Container 🤔 I think There are same functions in the eventDispatcher.ts

I thought we can just use like below code 🤔 
`this.eventDispatcher.addEventListener('renderAll', this.drawAll)`

Reply:
@mojosoeun 
I wrote it like this to use the following expression when registering an event listener in a class.
`containerRef.current addEventListener('renderAll', this.drawAll)`

After listening to you,
I think it would be better to inherit EventDispatch and use it :)